File: sources/Hera b1/dev/hera/client/utils/RenderUtils.java
package dev.hera.client.utils;

import org.lwjgl.opengl.GL11;

public class RenderUtils {

    public static void drawRoundedRect(double x, double y, double x1, double y1, double radius, int color) {
        float alpha = (color >> 24 & 255) / 255.0f;
        float red = (color >> 16 & 255) / 255.0f;
        float green = (color >> 8 & 255) / 255.0f;
        float blue = (color & 255) / 255.0f;
        GL11.glEnable(GL11.GL_BLEND);
        GL11.glDisable(GL11.GL_TEXTURE_2D);
        // GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
        GL11.glEnable(GL11.GL_LINE_SMOOTH);
        GL11.glPushMatrix();
        GL11.glColor4f(red, green, blue, alpha);
        x *= 2.0D;
        y *= 2.0D;
        x1 *= 2.0D;
        y1 *= 2.0D;
        GL11.glScaled(0.5D, 0.5D, 0.5D);
        GL11.glBegin(9);

        int i;
        for(i = 0; i <= 90; i += 3) {
            GL11.glVertex2d(x + radius + Math.sin((double)i * 3.141592653589793D / 180.0D) * radius * -1.0D, y + radius + Math.cos((double)i * 3.141592653589793D / 180.0D) * radius * -1.0D);
        }

        for(i = 90; i <= 180; i += 3) {
            GL11.glVertex2d(x + radius + Math.sin((double)i * 3.141592653589793D / 180.0D) * radius * -1.0D, y1 - radius + Math.cos((double)i * 3.141592653589793D / 180.0D) * radius * -1.0D);
        }

        for(i = 0; i <= 90; i += 3) {
            GL11.glVertex2d(x1 - radius + Math.sin((double)i * 3.141592653589793D / 180.0D) * radius, y1 - radius + Math.cos((double)i * 3.141592653589793D / 180.0D) * radius);
        }

        for(i = 90; i <= 180; i += 3) {
            GL11.glVertex2d(x1 - radius + Math.sin((double)i * 3.141592653589793D / 180.0D) * radius, y + radius + Math.cos((double)i * 3.141592653589793D / 180.0D) * radius);
        }


        GL11.glEnd();
        GL11.glScaled(2.0D, 2.0D, 2.0D);
        GL11.glEnable(GL11.GL_TEXTURE_2D);
        GL11.glDisable(GL11.GL_LINE_SMOOTH);
        GL11.glDisable(GL11.GL_BLEND);

        GL11.glPopMatrix();
        GL11.glColor3f(255, 255, 255);

    }

    public static void drawRoundedRect(float x, float y, float x1, float y1, float radius, int color) {
        drawRoundedRect((double)x, (double)y, (double)x1, (double)y1, (float)radius, color);
    }

    public static void drawRect(float paramXStart, float paramYStart, float paramXEnd, float paramYEnd, int paramColor) {
        float alpha = (float)(paramColor >> 24 & 255) / 255.0F;
        float red = (float)(paramColor >> 16 & 255) / 255.0F;
        float green = (float)(paramColor >> 8 & 255) / 255.0F;
        float blue = (float)(paramColor & 255) / 255.0F;
        GL11.glEnable(3042);
        GL11.glDisable(3553);
        GL11.glBlendFunc(770, 771);
        GL11.glEnable(2848);
        GL11.glPushMatrix();
        GL11.glColor4f(red, green, blue, alpha);
        GL11.glBegin(7);
        GL11.glVertex2d((double)paramXEnd, (double)paramYStart);
        GL11.glVertex2d((double)paramXStart, (double)paramYStart);
        GL11.glVertex2d((double)paramXStart, (double)paramYEnd);
        GL11.glVertex2d((double)paramXEnd, (double)paramYEnd);
        GL11.glEnd();
        GL11.glPopMatrix();
        GL11.glEnable(3553);
        GL11.glDisable(2848);
    }

    public static void drawRect(double x, double y, double x2, double y2, int color){
        drawRect((float)x, (float)y, (float) x2, (float)y2, color);
    }

    public static void drawBorder(float x, float y, float f, float g, int i, int rgb) {
        drawRect(x, y, f, y + (float)i, rgb);
        drawRect(x, g, f, g + (float)i, rgb);
        drawRect(x, y, x + (float)i, g + (float)i, rgb);
        drawRect(f, y, f + (float)i, g + (float)i, rgb);
    }

    public static void drawBorder(double x, double y, double f, double g, double i, int rgb) {
        drawRect(x, y, f, y + i, rgb);
        drawRect(x, g, f, g + i, rgb);
        drawRect(x, y, x + i, g + i, rgb);
        drawRect(f, y, f + i, g + i, rgb);
    }

}

File: sources/Leaf devEnv/com/leafclient/leaf/utils/render/RenderUtils.kt
package com.leafclient.leaf.utils.render

import com.leafclient.leaf.utils.math.Area
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.ScaledResolution
import net.minecraft.client.shader.Framebuffer
import org.lwjgl.BufferUtils
import org.lwjgl.opengl.GL11.*
import java.awt.Color
import kotlin.math.roundToInt


object RenderUtils {

    private val constants = intArrayOf(GL_ALPHA_TEST, GL_BLEND, GL_DEPTH_TEST, GL_LINE_SMOOTH, GL_POINT_SMOOTH, GL_TEXTURE_2D)

    private val values = BooleanArray(constants.size)
    private val color = BufferUtils.createFloatBuffer(16)

    /**
     * Prepares a 2d render by modifying few gl [constants] and storing them
     */
    fun render2D(render: () -> Unit) {
        for (i in values.indices) {
            values[i] = glGetBoolean(constants[i])
        }
        glGetFloat(GL_CURRENT_COLOR, color)

        glPushMatrix()
        glEnable(GL_BLEND)
        glDisable(GL_TEXTURE_2D)
        glEnable(GL_LINE_SMOOTH)
        glEnable(GL_POINT_SMOOTH)
        glEnable(GL_ALPHA_TEST)

        render()

        glPopMatrix()

        glColor4f(color.get(0), color.get(1), color.get(2), color.get(3))
        for (i in values.indices.reversed()) {
            val value = values[i]
            if (value) {
                glEnable(constants[i])
            } else {
                glDisable(constants[i])
            }
        }
    }

    /**
     * Draws [area] rectangle
     */
    fun rect(area: Area, color: Color) {
        render2D {
            glColor4f(color.red / 255F, color.green / 255F, color.blue / 255F, color.alpha / 255F)
            glBegin(GL_QUADS)
            glVertex2d(area.x.toDouble(), area.y.toDouble())
            glVertex2d(area.x.toDouble(), area.y.toDouble() + area.height.toDouble())
            glVertex2d(area.x.toDouble() + area.width.toDouble(), area.y.toDouble() + area.height.toDouble())
            glVertex2d(area.x.toDouble() + area.width.toDouble(), area.y.toDouble())
            glEnd()
        }
    }


    /**
     * Draws the [framebuffer] onto the screen.
     */
    fun drawFramebuffer(framebuffer: Framebuffer) {
        val scaledResolution = ScaledResolution(Minecraft.getMinecraft())
        glBindTexture(GL_TEXTURE_2D, framebuffer.framebufferTexture)
        glBegin(GL_QUADS)
        glTexCoord2d(0.0, 1.0)
        glVertex2d(0.0, 0.0)
        glTexCoord2d(0.0, 0.0)
        glVertex2d(0.0, scaledResolution.scaledHeight.toDouble())
        glTexCoord2d(1.0, 0.0)
        glVertex2d(scaledResolution.scaledWidth.toDouble(), scaledResolution.scaledHeight.toDouble())
        glTexCoord2d(1.0, 1.0)
        glVertex2d(scaledResolution.scaledWidth.toDouble(), 0.0)
        glEnd()
    }

    /**
     * Scissors [render]
     */
    fun scissor(area: Area, stopsScissoring: Boolean = true, render: () -> Unit)
        = scissor(area.x, area.y, area.width, area.height, stopsScissoring, render)

    /**
     * Scissors [render]
     */
    fun scissor(x: Float, y: Float, width: Float, height: Float, stopsScissoring: Boolean = true, render: () -> Unit) {
        val scaleFactor = 2.0F * Screen.screenScale

        glEnable(GL_SCISSOR_TEST)
        glScissor(
            (x * scaleFactor).roundToInt(),
            (Minecraft.getMinecraft().displayHeight - (y + height) * scaleFactor).roundToInt(),
            (width * scaleFactor).roundToInt(),
            (height * scaleFactor).roundToInt()
        )
        render()
        if(stopsScissoring)
            glDisable(GL_SCISSOR_TEST)
    }

    /**
     * Interpolates the rendering
     */
    fun interpolate(now: Double, then: Double, ticks: Float): Double {
        return then + (now - then) * ticks
    }

}